### PR TITLE
Add Smartcard HSM support and bug fixes

### PR DESCRIPTION
**Pull request by Manuthor**
The author of this PR is @SG-O.

### Motivation

Smartcard HSM-based devices (such as the Nitrokey HSM 2) are inexpensive, widely available, and provide strong hardware-backed key protection. Since they are supported by a PKCS#11 library, they can be integrated relatively easily into Cosmian KMS.

### This PR
This PR introduces Smartcard HSM support and a number of bug fixes and improvements to the overall HSM support in Cosmian KMS:
- Adds support for Smartcard HSM devices.
- Implements AES-CBC support (with PKCS#7 padding) so symmetric encryption is available on HSMs that do not support AES-GCM.
- Adds automatic algorithm selection logic for encryption/decryption on HSMs based on the HSM's actual capabilities.
- Adds mechanisms discovery (get_supported_mechanisms / get_mechanism_info) to test for these capabilities.
- Adds multi round AES-CBC encryption/decryption to work around HSM hardware limitations. This way large payloads can be encrypted on resource-constrained HSMs.
- Fixes bugs related to object id and object handle lookup.
- Makes the existing KMS HSM tests compatible with a broader range of HSMs and adds test helpers to pass all HSM test variables via environment variables.
- Adds test for all new functions and improves existing tests impacted by the changes without changing expected default behavior.

### Notes
Smartcard HSMs provide robust security but have some limitations (e.g., no AES GCM support, limited buffer size for encryption). These limitations are handled gracefully in this PR by:
- Automatically selecting the best algorithm possible (AES CBC in this case). This was also implemented for RSA encryption.
- Splitting data into manageable chunks during AES CBC encryption.
Further improvements could include a capability database for specific HSM models to fine-tune automatic algorithm selection and optimal chunk sizes.

### How to test these changes
Install SC HSM library as described here: https://github.com/CardContact/sc-hsm-embedded
Initialize the Smartcard HSM with default dev pin (648219)
Run HSM specific thest with 
```
HSM_USER_PASSWORD=648219 HSM_SLOT_ID=1 cargo test --target x86_64-unknown-linux-gnu --features smartcardhsm -- tests::test_hsm_all
```
Run KMS smartcardhsm tests with
```
HSM_MODEL=smartcardhsm;HSM_USER_PASSWORD=648219;HSM_SLOT_ID=1 test --features non-fips
```
You can now also run KMS softhsm2 tests with
```
HSM_MODEL=softhsm2;HSM_SLOT_ID=XXXXXXX;HSM_USER_PASSWORD=XXXXXXX test --features non-fips
```

### Suggested areas to focus review
base_hsm session_impl.rs: 
- object attribute usage (CKA_LABEL vs CKA_ID) in "get_object_id".
- the newly introduced handling of multiple matches in "find_object_handle"s.
- AES-CBC implementation: PKCS#7 padding correctness annd the chunking/rounding logic used to work around HSM limits.
- The algorithm used to determine the supported hashing algorithms for RSA OAEP implemented in "get_supported_oaep_hash".

base_hsm slots.rs:
- The implementation used to get the HSM's supported algorithms ("get_mechanism_info" and "get_supported_mechanisms").

base_hsm base_hsm.rs:
- The algorithm used in "get_algorithms" to match the reported capabilities to internal values.

interfaces/src/hsm encryption_oracle_impl.rs:
- Algorithm selection heuristics: ensure selection is safe and appropriate across HSMs. Also check fallback error usage and OAEP hash selection.

Tests: ensure the new and changed tests are correct and work with all possible devices.
Security considerations: confirm we do not leak sensitive material in logs or error messages.

### Backwards compatibility
These changes should be backward-compatible from the perspective of existing KMS deployments and existing CI testing:
The changes only add functionality without changing existing behavior (except fixing bugs for SoftHSM2)
No wire or on-disk format changes are introduced by these changes.
There are no required migrations.

### CODE REVIEW

**Comment by SG-O** [comment]
@Manuthor I added a way to provide HSM specific capabilities to the BaseHsm implementation and made usage of the multi round CBC implementation dependent on these capabilities.

**Comment by Manuthor** [comment]
> @Manuthor I added a way to provide HSM specific capabilities to the BaseHsm implementation and made usage of the multi round CBC implementation dependent on these capabilities.

Well done! 
Do you have the correct permissions to push on this PR your new commits?

**Comment by SG-O** [comment]
@Manuthor Could you merge my branch into the upstream one?

**Comment by SG-O** [comment]
@Manuthor Sorry, I messed up the rebase and reverted the change to the SoftHsm library path to the old value.
I fixed the issue and also applied formatting so the clippy test should pass.
Please merge.

File: crate/hsm/smartcardhsm/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+keywords.workspace = true
+categories.workspace = true
+description = "Smartcard HSM PKCS#11 loader"
+
+[lib]
+doctest = false
+
+[dependencies]
+cosmian_kms_base_hsm = { path = "../base_hsm", version = "5.5.1" }
+cosmian_kms_interfaces = { path = "../../interfaces", version = "5.5.1" }
```
Comment (Manuthor) [inline crate/hsm/smartcardhsm/Cargo.toml]: For publication reasons (cargo publish), this version must be aligned to last version (and will be bumped afterwards)
```suggestion
cosmian_kms_base_hsm = { path = "../base_hsm", version = "5.8.1" }
cosmian_kms_interfaces = { path = "../../interfaces", version = "5.8.1" }
```

File: crate/interfaces/src/hsm/encryption_oracle_impl.rs [hunk 2]
```diff
@@ -145,3 +147,42 @@ fn parse_uid(uid: &str) -> InterfaceResult<(usize, Vec<u8>)> {
     })?;
     Ok((slot_id, key_id.as_bytes().to_vec()))
 }
+
+/// Selects a default AES algorithm from the provided list of supported algorithms.
+///
+/// Preference order:
+/// 1. `AesGcm`
+/// 2. `AesCbc`
+fn parse_default_aes_algorithm(
```
Comment (Manuthor) [inline crate/interfaces/src/hsm/encryption_oracle_impl.rs]: I believe these 2 parsing functions should be handled directly on `CryptoAlgorithm` impl. I would declare your 2 new functions in `CryptoAlgorithm` impl (renaming them to `get_aes_algorithm` and `get_rsa_algorithm` )

File: crate/hsm/base_hsm/src/kms_hsm.rs [hunk 3]
```diff
@@ -38,10 +38,18 @@ use cosmian_kms_interfaces::{
...
 
 #[async_trait]
-impl HSM for BaseHsm {
+impl<P: HsmProvider> HSM for BaseHsm<P> {
+    async fn get_supported_algorithms(
+        &self,
+        slot_id: usize,
+    ) -> InterfaceResult<Vec<CryptoAlgorithm>> {
+        let supported = self.get_algorithms(slot_id)?;
+        Ok(supported)
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/kms_hsm.rs]:
```suggestion
        Ok(self.get_algorithms(slot_id)?)
```
unless you need to print it

File: crate/hsm/base_hsm/src/session/session_impl.rs [hunk 4]
```diff
@@ -200,19 +233,118 @@ impl Session {
...
+    /// # Errors
+    /// * Returns an error if RSA key pair generation fails.
+    /// * Returns an error if the HSM library does not provide the `C_EncryptInit` function.
+    /// * Returns an error if destroying the temporary test keys fails.
+    ///
+    /// # Safety
+    /// This function calls unsafe FFI functions from the HSM library. All temporary keys are
+    /// cleaned up after testing.
+    pub fn get_supported_oaep_hash(&self) -> HResult<Vec<CK_MECHANISM_TYPE>> {
+        let mut cache = self.supported_oaep_hash_cache.lock().unwrap();
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/session/session_impl.rs]: Unhandled unwrap here.

Ideally, to catch errors like this, we should (I'll do it later) enable the clippy-workspace lints for all subcrates and in particular the `base_hsm` subcrate, by uncommenting this part in Cargo.toml:

```
# [lints]
# workspace = true
```

Before fixing those clippy lints, we should at least avoid panic setting those rules (in `crate/hsm/base_hsm/Cargo.toml`):

```
[lints.clippy]
unwrap_used = "deny"
expect_used = "deny"
panic = "deny"
panic_in_result_fn = "deny"

```

File: crate/hsm/base_hsm/src/session/session_impl.rs [hunk 5]
```diff
@@ -200,19 +233,118 @@ impl Session {
...
-            pValue: object_id.as_ptr() as CK_VOID_PTR,
-            ulValueLen: object_id.len() as CK_ULONG,
-        }];
+        // Create a temporary key for testing
+        let sk_id = Uuid::new_v4().to_string();
+        let pk_id = sk_id.clone() + "_pk";
+        let (sk_handle, pk_handle) = self.generate_rsa_key_pair(
+            sk_id.as_bytes(),
+            pk_id.as_bytes(),
+            RsaKeySize::Rsa1024,
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/session/session_impl.rs]: RSA1024 being deprecated, add a comment ensuring that in this case size of RSA modulus does not matter here.

File: crate/hsm/base_hsm/src/session/session_impl.rs [hunk 6]
```diff
@@ -222,40 +354,109 @@ impl Session {
                 template.len() as CK_ULONG,
             );
             if rv != CKR_OK {
-                return Err(HError::Default(format!("C_FindObjectsInit failed: {rv}")));
+                return Err(HError::Default(
+                    "Failed to initialize object search".to_string(),
+                ));
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/session/session_impl.rs]:
```suggestion
                return Err(HError::Default(format!("Failed to initialize object search: C_FindObjectsInit failed: {rv}")));
```

File: crate/hsm/base_hsm/src/session/session_impl.rs [hunk 7]
```diff
@@ -222,40 +354,109 @@ impl Session {
...
-            );
-            if rv != CKR_OK {
-                return Err(HError::Default(format!("C_FindObjects failed: {rv}")));
+            loop {
+                let rv = self.hsm.C_FindObjects.ok_or_else(|| {
+                    HError::Default("C_FindObjects not available on library".to_string())
+                })?(
+                    self.session_handle,
+                    &raw mut object_handle,
+                    1,
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/session/session_impl.rs]:
```suggestion
                    1, // ulMaxObjectCount
```
And I wonder if this hard-limit should not be set in your `DefaultCapabilityProvider`.

File: crate/hsm/base_hsm/src/session/session_impl.rs [hunk 8]
```diff
@@ -222,40 +354,109 @@ impl Session {
...
+                    &raw mut object_count,
+                );
+                if rv != CKR_OK {
+                    return Err(HError::Default("Failed to find objects".to_string()));
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/session/session_impl.rs]:
```suggestion
                    return Err(HError::Default(format!("Failed to find objects: C_FindObjects failed: {rv}")));
```

File: crate/hsm/base_hsm/src/session/session_impl.rs [hunk 9]
```diff
@@ -222,40 +354,109 @@ impl Session {
...
+        let mut object_handles = self.find_object_handles(template.to_vec())?;
+        if object_handles.is_empty() {
+            if object_id_string.trim().ends_with("_pk") {
+                //Check if the HSM stores the object without the suffix
+                let trimmed = object_id_string.trim().strip_suffix("_pk");
+                let object_id_trimmed = match trimmed {
+                    Some(trimmed) => trimmed,
+                    None => object_id_string.trim(),
+                }
+                .as_bytes();
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/session/session_impl.rs]:
```suggestion
                let object_id_trimmed = object_id_string
                    .trim()
                    .strip_suffix("_pk")
                    .unwrap_or(object_id_string.trim())
                    .as_bytes();
```

File: crate/hsm/base_hsm/src/session/session_impl.rs (hunk 4 above)
Comment (Manuthor) [reply crate/hsm/base_hsm/src/session/session_impl.rs]: IMHO I would also add:
```
indexing_slicing = "deny"
```
But ignore it if you prefer.

File: crate/hsm/base_hsm/src/session/session_impl.rs [hunk 10]
```diff
@@ -222,40 +354,109 @@ impl Session {
...
+                ));
             }
+        }
+        Ok(object_handles)
+    }
+
+    pub fn get_object_handle(&self, object_id: &[u8]) -> HResult<CK_OBJECT_HANDLE> {
+        let object_id_string = String::from_utf8(Vec::from(object_id))
+            .map_err(|e| HError::Default(format!("Failed to convert object_id to string: {e}")))?;
+        debug!("Retrieving Object handle for id: {object_id_string}");
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/session/session_impl.rs]: I would remove this and use:
```
if object_id.ends_with(b"_pk") {
```
in place of:
```
if object_id_string.trim().ends_with("_pk") {
```

File: crate/hsm/base_hsm/src/session/session_impl.rs [hunk 11]
```diff
@@ -222,40 +354,109 @@ impl Session {
...
+        let template = [CK_ATTRIBUTE {
+            type_: CKA_LABEL,
+            pValue: object_id.as_ptr() as CK_VOID_PTR,
+            ulValueLen: object_id.len() as CK_ULONG,
+        }];
 
-            if object_count == 0 {
+        let mut object_handles = self.find_object_handles(template.to_vec())?;
+        if object_handles.is_empty() {
+            if object_id_string.trim().ends_with("_pk") {
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/session/session_impl.rs]: For my knowledge, why do we need here to get the PrivateKey handle? Could you add a comment please?

File: crate/hsm/base_hsm/src/session/session_impl.rs [hunk 12]
```diff
@@ -222,40 +354,109 @@ impl Session {
...
                 return Err(HError::Default("Object not found".to_string()));
             }
+        }
 
-            //update cache
-            self.object_handles_cache
-                .insert(object_id.to_vec(), object_handle);
-
-            Ok(object_handle)
+        let mut object_handle = object_handles[0];
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/session/session_impl.rs]: Handle the case where object_handles is empty.

File: crate/hsm/base_hsm/src/session/session_impl.rs [hunk 13]
```diff
@@ -393,6 +550,72 @@ impl Session {
...
+    /// PKCS#7 padding ensures that the input length is a multiple of the block size,
+    /// which is required for many block cipher encryption algorithms (such as AES in CBC mode).
+    ///
+    /// # Arguments
+    /// * `data` - The input data to be padded.
+    /// * `block_size` - The block size in bytes (commonly 16 for AES).
+    ///
+    /// # Returns
+    /// * `Vec<u8>` - A new buffer containing the original data with PKCS#7 padding appended.
+    fn pkcs7_pad(&self, data: Vec<u8>, block_size: usize) -> Vec<u8> {
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/session/session_impl.rs:601]: What about using the crate `block-padding` and something like this?
```
use block_padding::{Pkcs7, Padding};
use zeroize::Zeroizing;

fn pkcs7_pad(&self, data: Vec<u8>, block_size: usize) -> Vec<u8> {
    Pkcs7::pad(&data, block_size).to_vec()
}

fn pkcs7_unpad(
    &self,
    data: Zeroizing<Vec<u8>>,
    block_size: usize,
) -> HResult<Zeroizing<Vec<u8>>> {
    match Pkcs7::unpad(&data) {
        Ok(unpadded) => Ok(Zeroizing::new(unpadded.to_vec())),
        Err(_) => Err(HError::Default("Invalid PKCS#7 padding".to_string())),
    }
}
```
(To be verified)

File: crate/hsm/base_hsm/src/session/session_impl.rs [hunk 14]
```diff
@@ -559,6 +847,180 @@ impl Session {
...
+    /// the operation into multiple rounds if the input exceeds `max_round_length`.
+    /// This is useful for large data sets where encrypting in one call would exceed
+    /// the module's limits.
+    ///
+    /// Multiple rounds can be performed without compromising security because
+    /// each block of ciphertext becomes the initialization vector (IV) for the
+    /// next block. This function preserves that property by carrying forward the
+    /// final ciphertext block of one round as the IV for the next round. As a result,
+    /// the ciphertext produced by multi-round encryption is bit-for-bit identical
+    /// to what would be produced by a single-shot AES-CBC encryption with the same
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/session/session_impl.rs:921]: Do you think we could have a test to guaranty this assertion?

File: crate/hsm/base_hsm/src/session/session_impl.rs [hunk 15]
```diff
@@ -1214,6 +1677,17 @@ impl Session {
...
+            None => return Ok(Some(id)),
+            Some(key_type) => key_type,
+        };
+        let term = "_pk".as_bytes();
+        if id.ends_with(term) {
+            return Ok(Some(id))
+        }
+        if key_type == KeyType::RsaPublicKey {
+            id.append(&mut term.to_vec())
+        }
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/session/session_impl.rs]: Eventually try this:
```suggestion
        if let Some(KeyType::RsaPublicKey) = self.get_key_type(object_handle)? {
            if !id.ends_with(b"_pk") {
            id.extend_from_slice(b"_pk");
            }
        }
```

File: crate/hsm/smartcardhsm/README.md [hunk 16]
```diff
@@ -0,0 +1,21 @@
+# Smartcard HSM
```
Comment (Manuthor) [inline crate/hsm/smartcardhsm/README.md]: I believe the case is `SmartCard`

**Comment by Manuthor** [comment]
> @Manuthor Thanks for the invite.
> 
> The test failure in the macos_arm job looks like it has nothing to do with my changes. I ran it on my branch and there it passes. So maybe a race condition?
> 
> Anyways all other tests are now passing. The remaining issue is the lack of an emulator for CI testing. As a workaround, I could try to provide you with remote access to my hardware for running the tests. There is supposed to be a way to expose my SC HSM remotely over HTTP if that’s of interest, I can look into setting this up. That said, this wouldn’t be a solution for automated CI.
> 
> Once everything is in order, I'll of course go through all commits and do a clean rebase.

Indeed, it seems to be complicated to automate SC HSM tests in CI.

File: crate/hsm/base_hsm/src/session/session_impl.rs (hunk 9 above)
Comment (SG-O) [reply crate/hsm/base_hsm/src/session/session_impl.rs]: Due to the changes suggested in your next comment, this has been solved in a different way.

File: crate/hsm/base_hsm/src/session/session_impl.rs (hunk 13 above)
Comment (SG-O) [reply crate/hsm/base_hsm/src/session/session_impl.rs:601]: I thought about using an existing implementation. Not wanting to add random dependencies and the "clunky" usage of the implementation in "block-padding" (see https://docs.rs/block-padding/latest/block_padding/struct.Pkcs7.html) was the reason to implement this simple algorithm myself.

Of course I'm happy to change to "block-padding" if this is something you want me to do.

File: crate/hsm/base_hsm/src/session/session_impl.rs (hunk 14 above)
Comment (SG-O) [reply crate/hsm/base_hsm/src/session/session_impl.rs:921]: I have tests for this both in the Softsm2 tests as well as in the Smartcardhsm tests as part of "test_hsm_aes_cbc_encrypt" test. If you want, I'll put those tests in a separate function for clarity.

File: crate/hsm/base_hsm/src/session/session_impl.rs (hunk 11 above)
Comment (SG-O) [reply crate/hsm/base_hsm/src/session/session_impl.rs]: I hope I interpreted your comment correctly and you are talking about the "get_object_handle" function. I added documentation and a few comments to it. If you need more details or talking about something else, please let me know and I'll try to add clarifications.

File: crate/hsm/base_hsm/src/session/session_impl.rs (hunk 12 above)
Comment (SG-O) [reply crate/hsm/base_hsm/src/session/session_impl.rs]: I hope I interpreted your comment correctly and you are talking about the "get_object_handle" function.
The case is handled here:
https://github.com/Cosmian/kms/blob/5b1254961d5521dbf79f155f4b1b3edfcd6ea617/crate/hsm/base_hsm/src/session/session_impl.rs#L439-L458

File: crate/hsm/base_hsm/src/session/session_impl.rs (hunk 14 above)
Comment (SG-O) [reply crate/hsm/base_hsm/src/session/session_impl.rs:921]: I moved those tests into their own test function, improved its edge case coverage and made it more thorough over all.

File: crate/hsm/base_hsm/src/session/session_impl.rs (hunk 4 above)
Comment (SG-O) [reply crate/hsm/base_hsm/src/session/session_impl.rs]: I enabled these rules:
[lints.clippy]
unwrap_used = "deny"
expect_used = "deny"
panic = "deny"
panic_in_result_fn = "deny"
indexing_slicing = "deny"

and fixed all violations.

File: crate/server/src/tests/hsm/ec_dek.rs [hunk 17]
```diff
@@ -21,13 +21,16 @@ use crate::{
 };
 
 pub(super) async fn test_wrapped_ec_dek() -> KResult<()> {
-    let kek_uid = format!("hsm::0::{}", Uuid::new_v4());
+    let kek_uuid = Uuid::new_v4();
```
Comment (Manuthor) [inline crate/server/src/tests/hsm/ec_dek.rs:24]: I'm not sure about this because hsm::0 refers to slot 0 of a HSM.

File: crate/server/src/tests/hsm/rsa_dek.rs [hunk 18]
```diff
@@ -22,13 +22,16 @@ use crate::{
 };
 
 pub(super) async fn test_wrapped_rsa_dek() -> KResult<()> {
-    let kek_uid = format!("hsm::0::{}", Uuid::new_v4());
+    let kek_uuid = Uuid::new_v4();
```
Comment (Manuthor) [inline crate/server/src/tests/hsm/rsa_dek.rs:25]: Same here. Did you encounter difficulties with this syntax?

File: crate/server/src/tests/hsm/symmetric_dek.rs [hunk 19]
```diff
@@ -22,13 +22,16 @@ use crate::{
 };
 
 pub(super) async fn test_wrapped_symmetric_dek() -> KResult<()> {
-    let kek_uid = format!("hsm::0::{}", Uuid::new_v4());
+    let kek_uuid = Uuid::new_v4();
```
Comment (Manuthor) [inline crate/server/src/tests/hsm/symmetric_dek.rs:25]: Just for the memo

File: crate/hsm/base_hsm/src/base_hsm.rs [hunk 20]
```diff
@@ -85,6 +112,56 @@ impl BaseHsm {
...
+    /// # Errors
+    /// * Returns an error if the specified slot can't be accessed.
+    /// * Returns an error if the list of supported mechanisms can't be retrieved.
+    /// * Returns an error if the supported OAEP hashing algorithms can't be determined.
+    ///
+    /// # Safety
+    /// This function calls unsafe FFI functions from the HSM library to query mechanism information.
+    pub fn get_algorithms(&self, slot_id: usize) -> HResult<Vec<CryptoAlgorithm>> {
+        let slot = self.get_slot(slot_id)?;
+        let mechanisms: Vec<CK_MECHANISM_TYPE> = slot.get_supported_mechanisms()?;
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/base_hsm.rs]:
```suggestion
        let mechanisms = slot.get_supported_mechanisms()?;
```

File: crate/hsm/base_hsm/src/base_hsm.rs [hunk 21]
```diff
@@ -85,6 +112,56 @@ impl BaseHsm {
...
+        let session = slot.open_session(true)?;
+        let supported_hashes = session.get_supported_oaep_hash()?;
+        let mut algorithms: Vec<CryptoAlgorithm> = Vec::new();
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/base_hsm.rs]:
```suggestion
        let mut algorithms = Vec::new();
```

File: crate/hsm/base_hsm/src/base_hsm.rs [hunk 22]
```diff
@@ -3,28 +3,47 @@ use std::{
...
     fmt::{Display, Formatter},
+    marker::PhantomData,
     sync::{Arc, Mutex},
 };
 
-use pkcs11_sys::{CK_INFO, CKR_OK};
+use cosmian_kms_interfaces::CryptoAlgorithm;
+use cosmian_logger::debug;
+use pkcs11_sys::{
+    CK_INFO, CK_MECHANISM_TYPE, CKM_AES_CBC, CKM_AES_GCM, CKM_RSA_PKCS, CKM_RSA_PKCS_OAEP,
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/base_hsm.rs]:
```suggestion
    CK_INFO, CKM_AES_CBC, CKM_AES_GCM, CKM_RSA_PKCS, CKM_RSA_PKCS_OAEP,
```

File: crate/hsm/base_hsm/src/slots.rs [hunk 23]
```diff
@@ -108,32 +132,124 @@ impl SlotManager {
...
+                HError::Default("C_GetMechanismList not available on library".to_string())
+            })?(slot_id, ptr::null_mut(), &mut count);
+            if rv != CKR_OK {
+                return Err(HError::Default(
+                    "Failed to get mechanism count from HSM".to_string(),
+                ));
+            }
+
+            // Get mechanism list
+            let mut mechanisms: Vec<CK_MECHANISM_TYPE> = vec![0; count as usize];
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/slots.rs]:
```suggestion
            let mut mechanisms = vec![0; count as usize];
```

File: crate/hsm/base_hsm/src/session/session_impl.rs [hunk 24]
```diff
@@ -222,44 +358,163 @@ impl Session {
...
+                    self.session_handle,
+                    handles_buf.as_mut_ptr(),
+                    self.hsm_capabilities.find_max_object_count, //ulMaxObjectCount
+                    &raw mut object_count,
+                );
+                if rv != CKR_OK {
+                    return Err(HError::Default(format!(
+                        "Failed to find objects: C_FindObjects failed: {rv}"
+                    )));
+                }
```
Comment (Manuthor) [inline crate/hsm/base_hsm/src/session/session_impl.rs]: I'm pretty sure we can create a macro for this (and "systematize" the print of the return code everywhere in this file). I can do it later one.

File: crate/server/src/tests/hsm/ec_dek.rs (hunk 17 above)
Comment (SG-O) [reply crate/server/src/tests/hsm/ec_dek.rs:24]: hsm::0 referring to slot 0 was the issue I had to solve to make these tests compatible with HSMs other then utimaco. 
The kek_uuid now only refers to the key id. kek_uuid is passed to hsm_clap_config where it is combined with the appropriate slot to create the full kek uid:

https://github.com/Cosmian/kms/blob/b2d92f9a562527311d86a6dc414f9c351d734370/crate/server/src/tests/hsm/mod.rs#L87-L92

If you want, I can rename the variables to be more intuitive.

Comment (Manuthor) [reply crate/server/src/tests/hsm/ec_dek.rs:24]: That's good for me, thank you
